File: client/src/api/guest-portal.ts
import { api } from './client'

export interface GuestOrderItem {
  menu_item_id?: string
  name: string
  qty: number
  price_minor_units: number
}

export interface MenuItem {
  id: string
  name: string
  category: string
  price_minor_units: number
  prep_minutes: number
  is_available: boolean
}

export interface GuestOrder {
  id: string
  status: string
  items: GuestOrderItem[]
  total_minor_units: number
  received_at: string
  preparing_at: string | null
  delivering_at: string | null
  delivered_at: string | null
}

export interface GuestMaintenance {
  id: string
  description: string
  status: string
  urgency: string
  reported_at: string
  assigned_at: string | null
  resolved_at: string | null
}

export interface GuestCleaning {
  id: string
  priority: string
  preferred_time: string
  note: string | null
  status: string
  requested_at: string
  completed_at: string | null
}

export interface GuestDashboard {
  guest_name: string
  room_number: number
  floor: number
  orders: GuestOrder[]
  maintenance_requests: GuestMaintenance[]
  cleaning_requests: GuestCleaning[]
}

export const guestPortalApi = {
  dashboard: () =>
    api.get<GuestDashboard>('/reception/guest-portal/dashboard').then((r) => r.data),

  menu: () =>
    api.get<MenuItem[]>('/room-service/menu', { params: { available_only: true } }).then((r) => r.data),

  createOrder: (items: GuestOrderItem[]) =>
    api.post<GuestOrder>('/reception/guest-portal/orders', { items }).then((r) => r.data),

  createMaintenance: (description: string) =>
    api
      .post('/reception/guest-portal/maintenance', { description })
      .then((r) => r.data),

  createCleaning: (data: { priority: string; preferred_time: string; note?: string }) =>
    api
      .post<GuestCleaning>('/reception/guest-portal/cleaning', data)
      .then((r) => r.data),
}

File: server/reception-service/src/api/schemas/guest_portal.py
"""Pydantic schemas for the guest self-service portal."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class GuestOrderCreate(BaseModel):
    """Guest submits a food order from menu."""
    items: list[dict] = Field(
        ...,
        min_length=1,
        description='[{"menu_item_id": "uuid", "name": "...", "qty": 1, "price_minor_units": 500}]',
    )


class GuestMaintenanceCreate(BaseModel):
    """Guest reports a maintenance issue."""
    description: str = Field(..., min_length=5, max_length=500)


class GuestCleaningCreate(BaseModel):
    """Guest requests room cleaning."""
    priority: str = Field(default="normal", pattern=r"^(low|normal|high)$")
    preferred_time: str = Field(default="afternoon", pattern=r"^(morning|afternoon|evening|now)$")
    note: str | None = Field(default=None, max_length=200)


class GuestOrderOut(BaseModel):
    id: uuid.UUID
    status: str
    items: list[dict]
    total_minor_units: int
    received_at: datetime
    preparing_at: datetime | None = None
    delivering_at: datetime | None = None
    delivered_at: datetime | None = None


class GuestMaintenanceOut(BaseModel):
    id: uuid.UUID
    description: str
    status: str
    urgency: str
    reported_at: datetime
    assigned_at: datetime | None = None
    resolved_at: datetime | None = None


class GuestCleaningOut(BaseModel):
    id: uuid.UUID
    priority: str
    preferred_time: str
    note: str | None = None
    status: str
    requested_at: datetime
    completed_at: datetime | None = None


class GuestDashboardOut(BaseModel):
    guest_name: str
    room_number: int
    floor: int
    orders: list[GuestOrderOut]
    maintenance_requests: list[GuestMaintenanceOut]
    cleaning_requests: list[GuestCleaningOut]
